src/riak_kv_vnode.erl

```diff
@@ -400,27 +412,71 @@ tictac_rebuild(B, K, V) ->
     Clock = element(1, riak_object:summary_from_binary(V)),
     {IndexN, Clock}.
 
--spec rebuildtrees_workerfun(partition(), state()) -> fun().
 %% @doc
-%% A wrapper around the node_worker_pool to provide workers for tictac aae
-%% folds.  As the pool is a named pool started with riak_core, should be safe
-%% to used in init of inidivdual vnodes.
-rebuildtrees_workerfun(Partition, State) ->
-    fun(FoldFun, FinishFun) ->
-        Sender = self(),
-        ReturnFun = tictac_returnfun(Partition, trees),
-        FinishPlusReturnFun =
-            fun(WorkOutput) ->
-                FinishFun(WorkOutput),
-                ReturnFun(ok)
+%% Queue a tictac tree rebuild.  There are occasions when all vnodes queue this
+%% at the same time, so important that the snapshot for the rebuild is taken
+%% only when the fold is initiated.  Otherwise the snapshot may expire whilst
+%% sat on the queue
+-spec queue_tictactreerebuild(pid(), partition(), boolean(), state()) -> ok.
+queue_tictactreerebuild(AAECntrl, Partition, OnlyIfBroken, State) ->
+    Preflists = riak_kv_util:responsible_preflists(Partition),
+    Sender = self(),
+    ReturnFun = tictac_returnfun(Partition, trees),
+    FoldFun =
+        fun() ->
+            lager:info("Starting tree rebuild for partition=~w", [Partition]),
+            SW = os:timestamp(),
+            case when_loading_complete(AAECntrl,
+                                        Preflists,
+                                        fun preflistfun/2,
+                                        OnlyIfBroken) of
+                {ok, StoreFold, FinishFun} ->
+                    Output = StoreFold(),
+                    FinishFun(Output),
+                    Duration =
+                        timer:now_diff(os:timestamp(), SW) div (1000 * 1000),
+                    lager:info("Tree rebuild complete for partition=~w" ++
+                                " in duration=~w seconds", 
+                                [Partition, Duration]);
+                skipped ->
+                    lager:info("Tree rebuild skipped for partition=~w",
+                                [Partition])
             end,
-        Pool = select_queue(be_pool, State),
-        riak_core_vnode:queue_work(Pool, 
-                                    {fold, FoldFun, FinishPlusReturnFun},
-                                    Sender,
-                                    State#state.vnode_pool_pid)
+            ok
+        end,
+    JustReturnFun =
+        fun(ok) ->
```

Comment:
How sure can we be that this JustReturnFun is always called with 'ok' as an argument?
The way I read it is that this is the FinishFun on a foldfun... possibly read it wrong.

But if I read this right and the StoreFun is going to return something else then 'ok', then this finish fun crashes.

Reply:
https://github.com/basho/riak_kv/blob/develop-2.9/src/riak_kv_worker.erl#L46 

So JustReturnFun will always be passed the outcome of FoldFun.  But in this case the FoldFun must always return ok:

https://github.com/basho/riak_kv/blob/mas-i1765-tokenbucket/src/riak_kv_vnode.erl#L445